enterprise/server/cmd/ci_runner/main.go

```diff
@@ -1415,11 +1422,19 @@ func (ws *workspace) sync(ctx context.Context) error {
 	if err := git(ctx, ws.log, cleanArgs...); err != nil {
 		return err
 	}
-	// Create the branch if it doesn't already exist, then update it to point to
-	// the pushed branch tip.
-	if err := git(ctx, ws.log, "checkout", "--force", "-B", checkoutLocalBranchName, checkoutRef); err != nil {
+
+	// Checkout the git branch
+	if checkoutRef != "" {
+		if err := git(ctx, ws.log, "checkout", "--force", checkoutRef); err != nil {
+			return err
+		}
+	}
+
+	// Checkout a specific commit on the branch and create the local branch if it doesn't already exist
+	if err := git(ctx, ws.log, "checkout", "--force", "-B", checkoutLocalBranchName, checkoutCommit); err != nil {
```

Comment:
So when `checkoutRef != ""` that means `checkoutRef ~= pushedBranch || targetBranch`

In those cases, `checkoutLocalBranchName == pushedBranch || targetBranch`

And thus, `checkoutRef ~= checkoutLocalBranchName`

Should we simplify this then?

```suggestion
	// Checkout a specific commit on the branch and create the local branch if it doesn't already exist
	if err := git(ctx, ws.log, "checkout", "--force", "-B", checkoutLocalBranchName, checkoutCommit); err != nil {
```

---

Side note: 

We are using a sparse clone repo where objects are fetched remotely on git-checkout. So I am quite cautious in introducing new `git-checkout` calls. If we could make do with 1 checkout, it would be much preferably.

In the future, we might be able to make do without a sparse-clone by caching the git-bundle in our executor file cache(1).

(1): https://git-scm.com/docs/git-clone#Documentation/git-clone.txt---bundle-urilturigt

Reply:
I think at a high level, the changes we want to make are:

* `hostedrunner.go` should stop passing `target_commit_sha` and instead start passing both `commit_sha` (optional) and `target_branch` (required). This is because we currently don't have a way to fetch a commit sha from the remote (we only fetch branches currently), so I think that effectively means that `--target_branch` is always required. If you look at how `ws.fetch()` is called, it takes in a list of branches and is a NOP if no branches are passed in. So that means we don't fetch anything unless we have at least some branch name to go on.

Maybe @vadimberezniker can double-check my understanding here - there is a codepath in `hostedrunner.go` that only sets `--target_commit_sha` without setting `--pushed_branch` or `--target_branch`, but I am not sure if this codepath works, because it looks like the ci_runner only ever tries to fetch the pushed_branch or target_branch refs.

* When checking out, we can do the following:

  * If `commit_sha` is not set, then just `git checkout --force -B <target_branch>`
  * If `commit_sha` is set, then additionally reset the checked out branch to point to `commit_sha`, by just appending the commit SHA argument: `git checkout --force -B <target_branch> <commit_sha>`.

We can remove the logic that was setting the local branch name to "local", if my reasoning above is correct re. the branch name effectively being required.

Reply:
It works if we don't pass a `target_branch` because we `git clone` [here](https://github.com/buildbuddy-io/buildbuddy/blob/master/enterprise/server/cmd/ci_runner/main.go#L1329). So by default you stay on the master branch, even if you don't fetch any additional branches

Reply:
😅 We definitely need a cleaner way to specify valid combos of parameters

Reply:
@sluongng @bduffany I was thinking of an edge case where the same commit is on master and the feature branch, but with different commit history, so just checking out the commit is insufficient info to get the correct git tree

Ex. 
Master: (HEAD) C1 - C2 - C3
Feature:  (HEAD) C2 - C1 - C4 - C3

Request: Checkout feature branch commit C1

If we set checkoutRef = C1, we will checkout C1 on Master branch. This will include C2. However if we had checked out C1 on the Feature branch, it would not include C2 and would include C4.

This is why I thought we might need 2 check outs - the first to get on the correct branch, and the second to checkout the commit on that branch. Does my understanding of this match what you guys are thinking?

Reply:
I don't have a perfect understanding of Git, but I think the above situation is not possible. The git object commit (over which the sha is computed) includes the parent commit so even if you have the same exact change on two branches, the commit sha will be different due to having different parent shas.